runtime/lua/vim/fs.lua

```diff
@@ -334,6 +327,46 @@ function M.find(names, opts)
   return matches
 end
 
+--- Find the first parent directory containing a specific "marker", relative to a buffer's
+--- directory.
+---
+--- Example:
+---
+--- ```lua
+--- -- Find the root of a Python project
+--- vim.fs.findroot(0, {'pyproject.toml', 'setup.py' })
+---
+--- -- Find the root of a git repository
+--- vim.fs.findroot(0, '.git')
+---
+--- -- Find the parent directory containing any file with a .csproj extension
+--- vim.fs.findroot(0, function(name, path)
+---   return name:match('%.csproj$') ~= nil
+--- end)
+--- ```
+---
+--- @param bufnr integer Buffer number, or 0 for current buffer.
+--- @param marker (string|string[]|fun(name: string, path: string): boolean) A marker, or list
+---               of markers, to search for. If a function, the function is called for each
+---               evaluated item and should return true if {name} and {path} are a match.
+--- @return string? # Directory containing one of the given markers, or nil if no directory was
+--- found.
+function M.findroot(bufnr, marker)
+  assert(bufnr, 'missing required argument: bufnr')
+  assert(marker, 'missing required argument: marker')
+  local paths = M.find(marker, {
+    upward = true,
+    path = vim.fs.dirname(vim.api.nvim_buf_get_name(bufnr)),
```

Comment:
Just to make sure: this name doesn't have to be modified somehow (e.g., because of symlinks)?

I use
```lua
path = vim.fn.fnamemodify(args.file, ':p:h')
```
in my personal snippet.

Reply:
nvim_buf_get_name appears to (and is documented to) return the absolute path, which avoids the need for `:p:h`.